tools/setup

```diff
@@ -1,51 +1,98 @@
-
 NVM_VERSION="v0.34.0"
-
-BAZEL_DEB="https://github.com/bazelbuild/bazel/releases/download/0.29.0/bazel_0.29.0-linux-x86_64.deb"
-BAZEL_PKG_NAME=$(basename "$BAZEL_DEB")
+BAZEL_VERSION="0.29.0"
 
 YLW="\033[33m"
 RED="\033[91m"
 GRN="\033[92m"
 MAG="\033[95m"
 BLD="\033[1m"
 END="\033[0m"
-
 CMD="$BLD$YLW"
 
+warn() {
+    echo "$RED$BLD$1$END"
+}
 
 fail() {
-    echo "$RED$BLD$1$END"
+    warn $1
     exit 1
 }
 
 status() {
     echo "$MAG$1$END"
 }
 
+# Detect Platform
+case $(uname | tr '[:upper:]' '[:lower:]') in
+  linux*)
+    OS_NAME=linux
+    BAZEL_DEP="https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel_$BAZEL_VERSION-installer-linux-x86_64.sh"
+    ;;
+  darwin*)
+    OS_NAME=osx
+    BAZEL_DEP="https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-darwin-x86_64.sh"
+    ;;
+  msys*)
+    OS_NAME=windows
+    BAZEL_DEP="https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-windows-x86_64.exe"
+    # TODO(alxr) investigate windows compatability
+    fail "Windows is currently unsupported. Please follow install instructions in the README."
+    ;;
+  *)
+    OS_NAME=notset
+    warn "Can't detect host OS"
+    ;;
+esac
+
+BAZEL_PKG_NAME=$(basename "$BAZEL_DEP")
+
+status "0. Pre-requisites"
+if [ "$OS_NAME" == "linux" ]; then
+  status "0.1 Installing pre-requisites for Bazel"
+  sudo apt-get install pkg-config zip g++ zlib1g-dev unzip python3
+elif [ "$OS_NAME" == "osx" ]; then
+  status "0.1. Checking for Command Line Tools"
+  xcode-select --version || xcode-select --install || fail "Unable to install Command Line Tools"
+fi
+
 status "1. Install nvm"
-curl -o- "https://raw.githubusercontent.com/creationix/nvm/$NVM_VERSION/install.sh" | sh || \
+curl -L -o- "https://raw.githubusercontent.com/creationix/nvm/$NVM_VERSION/install.sh" | sh || \
     fail "Failed to download nvm version $NVM_VERSION."
 
 status "1.1 ensure nvm is in your current process"
 . ~/.nvm/nvm.sh || source ~/.nvm/nvm.sh
 
 status "2. Install node"
-(cd $ROOT && nvm install && nvm use) || fail 'Failed to install target node version.'
+pushd .
+cd $ROOT && nvm install && nvm use || fail 'Failed to install target node version.'
+popd
 
 status "3. Update npm to latest version"
 nvm install-latest-npm  || fail 'Failed to update npm to latest version.'
 
 status "4. Install dependencies"
 (cd $ROOT && npm ci && npm --prefix=server ci) || fail 'Failed to install dependencies.'
 
-status "5. Install bazel"
-(wget "$BAZEL_DEB" && set -x && sudo dpkg -i $BAZEL_PKG_NAME)
+status "4.1. Install sigh dependencies"
+(cd $ROOT && ./tools/sigh install) || fail 'Failed to install sigh dependencies.'
+
+status "5. Install Bazel"
+
+status "5.1. Downloading install script"
+curl -O -L "$BAZEL_DEP" || fail "Failed to download Bazel binary"
+
+status "5.2. Running install script"
+(chmod +x $BAZEL_PKG_NAME && ( . $BAZEL_PKG_NAME --user || source $BAZEL_PKG_NAME --user)) \
```

Comment:
You're sourcing the bazel install script here. Shouldn't you just run it instead?

e.g. from https://docs.bazel.build/versions/master/install-ubuntu.html#step-3-run-the-installer:
```
./bazel-<version>-installer-linux-x86_64.sh --user
```

Also, removing the `--user` bit should install it system-wide, which I think means you can get rid of the PATH stuff below.

Reply:
I see two issues here, let's talk about them separately: 

1. **source vs execute** I intended to use source so that the changes would take effect in the current shell environment. I'm not exactly sure what the shell script does, and I wanted to make sure that any updates to the env would be reflected by running the setup script. I do something like this in step 1.1.
2. **user vs system install** Out of habit I try to effect the minimal the scope of the dependency, for instance, preferring a dev install instead of a global install. The `--user` flag will treat Bazel like NVM moreso than treating it like apt-get. That's somehow more palatable to me -- what do you think?

Reply:
Sounds reasonable to me. I tend to prefer installing things globally rather than having scripts install things in random locations in my home directory and mess with my PATH, but that's just my personal preference. What you have here is totally fine :-)